pandas/core/arrays/categorical.py

```diff
@@ -578,7 +578,11 @@ def from_codes(cls, codes, categories, ordered=False):
             unordered.
         """
         try:
-            codes = np.asarray(codes, np.int64)
+            if (type(codes) == np.ndarray
```

Comment:
I wonder, is it sufficient to call `coerce_indexer_dtype(codes, categories)`? If so, how's the performance of that?

Reply:
```
arr = np.ones(10000000, 'int8')
# patch
In [4]: %timeit pd.Categorical.from_codes(arr, ['foo', 'bar'])
9.13 ms ± 104 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# coerce_indexer_dtype
In [4]: %timeit pd.Categorical.from_codes(arr, ['foo', 'bar'])
9.14 ms ± 134 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
But for lists,
```
arr = [1 for x in range(10000000)]
# patch
In [4]: %timeit pd.Categorical.from_codes(arr, ['foo', 'bar'])
600 ms ± 14.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# coerce_indexer_dtype
In [4]: %timeit pd.Categorical.from_codes(arr, ['foo', 'bar'])
2.32 s ± 23.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
I don't think `coerce_indexer_dtype` turns `codes` into an Numpy array.
```
diff --git a/pandas/core/arrays/categorical.py b/pandas/core/arrays/categorical.py
index 93e7aa7ca..8a096e4c8 100644
--- a/pandas/core/arrays/categorical.py
+++ b/pandas/core/arrays/categorical.py
@@ -578,18 +578,14 @@ class Categorical(ExtensionArray, PandasObject):
             unordered.
         """
         try:
-            if (type(codes) == np.ndarray
-                    and np.issubdtype(codes.dtype, np.integer)):
-                codes = np.asarray(codes)
-            else:
-                codes = np.asarray(codes, np.int64)
+            coerce_indexer_dtype(codes, categories)
         except (ValueError, TypeError):
             raise ValueError(
                 "codes need to be convertible to an arrays of integers")

         categories = CategoricalDtype.validate_categories(categories)

-        if len(codes) and (codes.max() >= len(categories) or codes.min() < -1):
+        if len(codes) and (np.max(codes) >= len(categories) or np.min(codes) < -1):
             raise ValueError("codes need to be between -1 and "
                              "len(categories)-1")
```

Reply:
Indeed. And if you combine the two approaches like 

`codes = coerce_indexer_dtype(np.asarray(codes), categories)`

does that work?

Reply:
```
In [3]: arr = np.ones(10000000, 'int8')
# patch
In [4]: %timeit pd.Categorical.from_codes(arr, ['foo', 'bar'])
9.06 ms ± 28.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# coerce_indexer_dtype + np.asarray
In [4]: %timeit pd.Categorical.from_codes(arr, ['foo', 'bar'])
9.03 ms ± 37.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [5]: lst = [1 for x in range(10000000)]
# patch
In [6]: %timeit pd.Categorical.from_codes(arr, ['foo', 'bar'])
588 ms ± 3.69 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# coerce_indexer_dtype + np.asarray
In [6]: %timeit pd.Categorical.from_codes(arr, ['foo', 'bar'])
583 ms ± 3.26 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Thoughts on that implementation? I slightly prefer it to the `if / else` version.

Reply:
I prefer it as well. I pushed a commit with your implementation.